rpcserver.go

```diff
@@ -1453,6 +1453,29 @@ func (r *rpcServer) ConnectPeer(ctx context.Context,
 	rpcsLog.Debugf("[connectpeer] requested connection to %x@%s",
 		peerAddr.IdentityKey.SerializeCompressed(), peerAddr.Address)
 
+	// Check if the connection timeout is set. If set, we will use it in our
+	// request.
+	if in.Timeout != "" && in.Timeout != "0s" {
+		timeout, err := time.ParseDuration(in.Timeout)
+		if err != nil {
+			rpcsLog.Errorf(
+				"[connectpeer]: error connecting to peer: %v", err,
+			)
+			return nil, err
+		}
+
+		// Set the new timeout value in our net. This will change the timeout
+		// value in our dialer, so we need to restore the system setting once
+		// the call is finished.
+		r.server.cfg.net.SetTimeout(timeout)
```

Comment:
Is it possible for this to race if we have two connect request with different values at the same time? Eg, request 1 sets value =30s, request 2 sets value = 40s, request 1 completes and replaces with default, request 2 then uses default? 

Another approach would be to add dial options to `Dial` in `tor.Net/Net` interface, so that we can provide these values on a per-call basis.

Reply:
I agree, forgot about checking race condition... Did try with the `Dial` approach though, turned out it required lots of code change. Originally I was thinking only allowing a global timeout so that it's easier to implement. Once that's done I thought, maybe users want to have some connections using a different timeout value, so I took this shortcut as it required minimal code change. Now that there's the race condition, I'll go back to modifying the function `Dial`. Thanks for pointing out!